llvm/tools/sycl-post-link/ModuleSplitter.cpp

```diff
@@ -211,43 +213,118 @@ EntryPointGroupVec groupEntryPointsByScope(const Module &M,
 
       Attribute Id = F.getFnAttribute(ATTR_SYCL_MODULE_ID);
       StringRef Val = Id.getValueAsString();
-      EntryPointMap[Val].push_back(&F);
+      EntryPointMap[Val].insert(&F);
       break;
     }
 
     case Scope_Global:
       // the map key is not significant here
-      EntryPointMap[GLOBAL_SCOPE_NAME].push_back(&F);
+      EntryPointMap[GLOBAL_SCOPE_NAME].insert(&F);
       break;
     }
   }
 
   if (!EntryPointMap.empty()) {
     EntryPointGroups.reserve(EntryPointMap.size());
-    for (auto &EPG : EntryPointMap)
-      EntryPointGroups.push_back({EPG.first, std::move(EPG.second)});
+    for (auto &EPG : EntryPointMap) {
+      EntryPointGroups.emplace_back(EntryPointGroup{
+          EPG.first, std::move(EPG.second), MD.getEntryPointGroup().Props});
+      EntryPointGroup &G = EntryPointGroups.back();
+      G.Props.Scope = EntryScope;
+    }
+  } else {
+    // No entry points met, record this.
+    EntryPointGroups.emplace_back(
+        EntryPointGroup{GLOBAL_SCOPE_NAME, EntryPointSet{}});
+  }
+  return EntryPointGroups;
+}
+
+template <class EntryPoinGroupFunc>
+EntryPointGroupVec
+groupEntryPointsByAttribute(const ModuleDesc &MD, StringRef AttrName,
+                            bool EmitOnlyKernelsAsEntryPoints,
+                            EntryPoinGroupFunc F) {
+  EntryPointGroupVec EntryPointGroups{};
+  std::map<StringRef, EntryPointSet> EntryPointMap;
+  const Module &M = MD.getModule();
+
+  // Only process module entry points:
+  for (const auto &F : M.functions()) {
+    if (!isEntryPoint(F, EmitOnlyKernelsAsEntryPoints) ||
+        !MD.isEntryPointCandidate(F)) {
+      continue;
+    }
+    if (F.hasFnAttribute(AttrName)) {
+      EntryPointMap[AttrName].insert(&F);
+    } else {
+      EntryPointMap[""].insert(&F);
+    }
+  }
+  if (!EntryPointMap.empty()) {
+    EntryPointGroups.reserve(EntryPointMap.size());
+    for (auto &EPG : EntryPointMap) {
+      EntryPointGroups.emplace_back(EntryPointGroup{
+          EPG.first, std::move(EPG.second), MD.getEntryPointGroup().Props});
+      F(EntryPointGroups.back());
+    }
   } else {
     // No entry points met, record this.
     EntryPointGroups.push_back({GLOBAL_SCOPE_NAME, {}});
+    F(EntryPointGroups.back());
   }
   return EntryPointGroups;
 }
 
+// Represents a call graph between functions in a module. Nodes are functions,
+// edges are "calls" relation.
+class CallGraph {
+public:
+  using FunctionSet = SmallPtrSet<const Function *, 16>;
+
+private:
+  std::unordered_map<const Function *, FunctionSet> Graph;
+  SmallPtrSet<const Function *, 1> EmptySet;
+
+public:
+  CallGraph(const Module &M) {
+    for (const auto &F : M) {
+      for (const Value *U : F.users()) {
+        if (const auto *I = dyn_cast<CallInst>(U)) {
+          if (I->getCalledFunction() == &F) {
```

Comment:
I think this can be an 'assert'.

Reply:
Can be addressed later.

Reply:
> I think this can be an 'assert'.

Actually no. consider `foo(a, b, &myfunc);` vs `myfunc(a, b);`